### ISSUE TITLE
Feature Request: Implement Automatic Differentiation (Autograd)

### ISSUE THREAD

**Issue by ajkdrag** [ajkdrag/bolt-rs#40]
# Feature Request: Implement Automatic Differentiation (Autograd)

## Proposed feature

This feature proposes the addition of a new crate, `bolt-autodiff`, to the `bolt-rs` framework. This crate will provide automatic differentiation (autograd) capabilities, enabling the computation of gradients for tensor operations. The core idea is to implement a "backend wrapper" pattern, where `bolt-autodiff` wraps existing `bolt-core` backends (like `bolt-cpu`) to intercept operations and record them in a computation graph.

This new crate will introduce:
- A `AutodiffBackend<B>` struct that implements the `bolt-core::backend::Backend` trait.
- A mechanism to dynamically build a computation graph (often referred to as a "tape") during the forward pass of tensor operations.
- The ability to perform a backward pass on this graph to compute gradients with respect to input tensors.
- An API for users to mark tensors as requiring gradients and to retrieve computed gradients.

## Motivation

Integrating autograd is a crucial step towards transforming `bolt-rs` into a comprehensive framework for machine learning and scientific computing. Without autograd, users must manually compute and implement gradients for optimization algorithms, which is error-prone, time-consuming, and severely limits the types of models and research that can be conducted with `bolt-rs`.

Key motivations include:
-   **Enabling Neural Network Training:** Autograd is fundamental for training neural networks efficiently. It removes the burden of manual gradient computation for complex architectures.
-   **Gradient-based Optimization:** Facilitates the use of optimization algorithms (e.g., Stochastic Gradient Descent, Adam) for a wide range of problems beyond neural networks.
-   **Research and Prototyping:** Accelerates the development and experimentation with new models and algorithms by simplifying the gradient computation aspect.
-   **Alignment with Modern ML Frameworks:** Brings `bolt-rs` in line with capabilities offered by frameworks like PyTorch, TensorFlow, and other Rust ML libraries like Burn and Candle, making it more appealing and powerful for users.

## Use case

A primary use case for this feature would be training a simple neural network. Consider a scenario where a user wants to implement and train a basic linear regression model or a multi-layer perceptron.

### Pseudo-code Example:

```rust
use std::sync::Arc;
use bolt_core::tensor::Tensor;
use bolt_cpu::{CpuBackend, CpuDevice};
// Assuming bolt_autodiff is available and provides AutodiffBackend
use bolt_autodiff::AutodiffBackend;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let device = CpuDevice::new();
    let cpu_backend = CpuBackend::new(device);
    let ad_backend = Arc::new(AutodiffBackend::new(cpu_backend));

    // Input data
    let x = Tensor::from_slice(&ad_backend, &[1.0, 2.0, 3.0, 4.0], &[4, 1])?.requires_grad(true);
    // Target data
    let y = Tensor::from_slice(&ad_backend, &[2.0, 4.0, 5.0, 4.0], &[4, 1])?;

    // Model parameters, requiring gradients
    let w = Tensor::from_slice(&ad_backend, &[0.0], &[1, 1])?.requires_grad(true); // Weight
    let b = Tensor::from_slice(&ad_backend, &[0.0], &[1])?.requires_grad(true);   // Bias

    let learning_rate = 0.01;
    let num_epochs = 100;

    for epoch in 0..num_epochs {
        // Forward pass
        let pred = w.matmul(&x)?.add(&b)?; // y_hat = Wx + b
        let loss = pred.sub(&y)?.powf(2.0)?.mean()?; // MSE loss

        // Zero out previous gradients (implementation detail handled by AutodiffBackend)
        // ad_backend.zero_grad(); // Or similar global mechanism

        // Backward pass
        loss.backward()?;

        // Update parameters using gradients
        // This part would ideally be handled by an optimizer, but shown here for clarity
        // w.data -= learning_rate * w.grad;
        // b.data -= learning_rate * b.grad;

        println!("Epoch {}: Loss = {}", epoch, loss.item()?);
    }

    // After training, inspect final parameters
    println!("Final Weight: {:?}", w.item()?);
    println!("Final Bias: {:?}", b.item()?);

    Ok(())
}
```

## Alternatives

During the initial research, several autograd implementation patterns were considered:

1.  **Compile-time Graph (e.g., `dfdx`):** This approach uses Rust's type system and const generics to build computation graphs at compile time. While offering strong guarantees and potential performance benefits, it can be less flexible for dynamic models where shapes or operations might change at runtime. Given `bolt-rs`'s current design with dynamic shapes (via `Layout`), this approach would require significant architectural changes.
2.  **Tape Wrapper (e.g., `candle`):** `candle` directly wraps its `Tensor` with a tape mechanism. This is similar to the proposed approach in spirit but might involve more direct modification or tight coupling within the `Tensor` structure itself.
3.  **Direct Integration into `bolt-core`:** Embedding autograd logic directly into `bolt-core`'s `Tensor` and `Backend` traits. This was rejected due to concerns about bloating the core library with functionality that might not always be needed and increasing its complexity. The backend wrapper approach keeps `bolt-core` lean and focused.

The chosen "backend wrapper" pattern (similar to `burn`'s `AutodiffBackend`) provides a good balance of flexibility, performance potential, and maintainability, allowing `bolt-core` to remain clean while enabling powerful autograd capabilities.

## Notes

The implementation will require careful consideration of Rust's ownership and borrowing rules, particularly for managing the mutable state of the computation graph and gradients (e.g., using `Arc<RwLock<...>>`). The design should also prioritize clear error handling and performance.

**Comment by ajkdrag** [comment]
Closed by #43
